set.c

```diff
@@ -0,0 +1,2149 @@
+        });
+        rb_gc_writebarrier_remember(ary);
+        rb_ary_set_len(ary, size);
+    }
+    else {
+        set_iter(set, set_to_a_i, (st_data_t)ary);
+    }
+    return ary;
+}
+
+/*
+ *  call-seq:
+ *    to_set(klass = Set, *args, &block) -> self or new_set
+ *
+ *  Returns self if receiver is an instance of +Set+ and no arguments or
+ *  block are given.  Otherwise, converts the set to another with
+ *  <tt>klass.new(self, *args, &block)</tt>.
+ *
+ *  In subclasses, returns `klass.new(self, *args, &block)` unless overridden.
+ */
+static VALUE
+set_i_to_set(int argc, VALUE *argv, VALUE set)
+{
+    VALUE klass;
+
+    if (argc == 0) {
+        klass = rb_cSet;
+        argv = &set;
+        argc = 1;
+    }
+    else {
+        klass = argv[0];
+        argv[0] = set;
+    }
+
+    if (klass == rb_cSet && rb_obj_is_instance_of(set, rb_cSet) &&
+            argc == 1 && !rb_block_given_p()) {
+        return set;
+    }
+
+    return rb_funcall_passing_block(klass, id_new, argc, argv);
+}
+
+/*
+ *  call-seq:
+ *    join(separator=nil)-> new_string
+ *
+ *  Returns a string created by converting each element of the set to a string.
+ */
+static VALUE
+set_i_join(int argc, VALUE *argv, VALUE set)
+{
+    rb_check_arity(argc, 0, 1);
+    return rb_ary_join(set_i_to_a(set), argc == 0 ? Qnil : argv[0]);
+}
+
+/*
+ *  call-seq:
+ *    add(obj) -> self
+ *
+ *  Adds the given object to the set and returns self.  Use `merge` to
+ *  add many elements at once.
+ *
+ *    Set[1, 2].add(3)                    #=> #<Set: {1, 2, 3}>
+ *    Set[1, 2].add([3, 4])               #=> #<Set: {1, 2, [3, 4]}>
+ *    Set[1, 2].add(2)                    #=> #<Set: {1, 2}>
+ */
+static VALUE
+set_i_add(VALUE set, VALUE item)
+{
+    rb_check_frozen(set);
+    if (set_iterating_p(set)) {
+        if (!set_lookup(RSET_TABLE(set), (st_data_t)item)) {
+            no_new_item();
+        }
+    }
+    else {
+        set_table *table = RSET_TABLE(set);
+        item = set_stringify(table, item);
+        RB_OBJ_WRITTEN(set, Qundef, item);
+        set_insert(table, (st_data_t)item);
+    }
+    return set;
+}
+
+/*
+ *  call-seq:
+ *    add?(obj) -> self or nil
+ *
+ *  Adds the given object to the set and returns self. If the object is
+ *  already in the set, returns nil.
+ *
+ *    Set[1, 2].add?(3)                    #=> #<Set: {1, 2, 3}>
+ *    Set[1, 2].add?([3, 4])               #=> #<Set: {1, 2, [3, 4]}>
+ *    Set[1, 2].add?(2)                    #=> nil
+ */
+static VALUE
+set_i_add_p(VALUE set, VALUE item)
+{
+    rb_check_frozen(set);
+    if (set_iterating_p(set)) {
+        if (!set_lookup(RSET_TABLE(set), (st_data_t)item)) {
+            no_new_item();
+        }
+        return Qnil;
+    }
+    else {
+        set_table *table = RSET_TABLE(set);
+        item = set_stringify(table, item);
+        RB_OBJ_WRITTEN(set, Qundef, item);
+        return set_insert(table, (st_data_t)item) ? Qnil : set;
```

Comment:
```suggestion
        if (set_insert(table, (st_data_t)item)) {
            return Qnil
        }
        RB_OBJ_WRITTEN(set, Qundef, item);
        return set;
```

We shouldn't trigger the write barrier if we didn't insert the item.